src/CMakeLists.txt

```diff
@@ -9,6 +10,18 @@ file(GLOB INL ../include/*.inl *.inl)
 file(GLOB_RECURSE EXTERNAL_INC ../external/*.h ../external/*.hpp)
 
 add_library(d3d11on12 SHARED ${SRC} ${INC} ${INL} ${EXTERNAL_INC} d3d11on12.rc d3d11on12.def)
+
+#if we don't have access to the d3d12translationlayer_wdk project, pull it from github (it's a part of the d3d12translationlayer repo)
+if(NOT TARGET d3d12translationlayer_wdk)

```

Comment:
I was also wondering if we should just change behavior here, and *always* use `FetchContent`. If we want to share a single `D3D12TranslationLayer` among multiple mapping layers, that'll automatically happen because `FetchContent` is cached at the top-level build directory, rather than per-sub-build. If we want to override `origin/master` (or a frozen SHA), a super-project just needs its own `FetchContent_Declare`.

I'm okay making this change later (or not at all, I guess), just wanted to throw it out there as a possibility.

Reply:
I was wondering what your thoughts would be on this, was having similar thoughts myself. I'm wondering if we should still have the `if` though, but instead of checking for the presence of `d3d12translationlayer_wdk` we check for a define/variable that a user could set in a parent CMakeLists. This is mainly to allow situations where you want to work on both projects without CMake spewing errors about creating duplicate targets.

Reply:
If we update *all* projects to use `FetchContent`, then the logic would be:
* Project A (I guess you meant D3D9On12 with your "both" comment?) uses `FetchContent_MakeAvailable`, pulls the code, and adds the project.
* Project B (D3D11On12) uses `FetchContent_MakeAvailable`, which implicitly sees that the `D3D12TranslationLayer` project has already been made available, and doesn't do it again.

Using a guard like this does enable other, non-`FetchContent` approaches to add the code, but given that `FetchContent` supports both superprojects overriding the declaration to local sources, or command line overriding to local sources, I'm thinking let's go all-in on it. It seems really useful/powerful.

Reply:
"Both" was referring to 11on12 and the translationLayer projects. Since the translation layer wouldn't be in the `_deps` folder at that time CMake pulls it again and then complains about the targets already existing. 
You can see in this snippet that the translation layer was already loaded ("Adding WDK-dependent project") but FetchContent still pulls the translation layer from git hub and tries to add its targets.

```
...
Adding WDK-dependent project.
Fetching D3D12TranslationLayer from GitHub...
CMake Error at C:/d3d/Build/_deps/d3d12translationlayer-src/DxbcParser/CMakeLists.txt:6 (add_library):
  add_library cannot create target "dxbcparser" because another target with
  the same name already exists.  The existing target is a static library
  created in source directory
  "C:/d3d/Direct3DMappingLayers/D3D12TranslationLayer/DxbcParser".  See
  documentation for policy CMP0002 for more details.
...
```

Reply:
For just working on projects that depend on the translation layer though Fetch Content would work great like you suggest, so I think we just need a way to also allow a user to say that they want to use their local copy of the translation layer and not fetch the one from git hub

Reply:
From the docs:
```
FETCHCONTENT_SOURCE_DIR_<ucName>
If this is set, no download or update steps are performed for the specified content and the <lcName>_SOURCE_DIR variable returned to the caller is pointed at this location. This gives developers a way to have a separate checkout of the content that they can modify freely without interference from the build. The build simply uses that existing source, but it still defines <lcName>_BINARY_DIR to point inside its own build area. Developers are strongly encouraged to use this mechanism rather than editing the sources populated in the default location, as changes to sources in the default location can be lost when content population details are changed by the project.
```

You can either set `FETCHCONTENT_SOURCE_DIR_D3D12TRANSLATIONLAYER` to an already-populated local source on the command line, or you can re-declare `FetchContent_Declare` using `SOURCE_DIR` as described on the `ExternalProject` docs:
```
Download Step Options:
A download method can be omitted if the SOURCE_DIR option is used to point to an existing non-empty directory.
```

So this should be okay, as long as nobody tries to both explicitly `add_subdirectory` into it *and* use a `FetchContent` dependency. As long all consumers are using `FetchContent`, it'll only be added once, and we can control whether the source is downloaded or provided locally.

Reply:
Oh very cool, I like that